### PR TITLE
use existing library function to lookup previous name hints

### PR DESCRIPTION
**Pull request by mjkkirschner**
### Purpose

I found that this PR:
https://github.com/DynamoDS/Dynamo/pull/9306

did indeed get migrations working using the migration.xml and AKA attributes - but not consistently with xml migrations.

This PR uses an existing function in library services to first check for an exact match using the full function signature (including parameter types) - and then falls back to checking for shortNames without parameters. This is more consistent with xml migration.

I've manually tested changing the name of a ZT node in the DSOffice namespace, adding that new function(without parameters) to the DSOffice.migrations.xml and all works as expected.

### TODO:
- [x] add a test for this migration
- [x] investigate how old tests pass, ~~likely they are not precise enough.~~ The tests did not include a case of migrating a type with parameters using a migration name hint that did not include those parameters - modified the test migration xml to include this case.

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [x] This PR modifies some build requirements and the readme is updated

### CODE REVIEW

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 1]
```diff
@@ -230,15 +230,13 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
             else if (typeof(DSFunctionBase).IsAssignableFrom(type))
             {
                 var mangledName = obj["FunctionSignature"].Value<string>();
-                var priorNames = libraryServices.GetPriorNames();
-                var functionDescriptor = libraryServices.GetFunctionDescriptor(mangledName);
-                string newName;
-
-                // Update the function descriptor if a newer migrated version of the node exists
-                if (priorNames.TryGetValue(mangledName, out newName))
+                var hintSignature = libraryServices.FunctionSignatureFromFunctionSignatureHint(mangledName);
```
Comment (aparajit-pratap) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs:233]: How about renaming this existing method to `GetFunctionSignatureFromFunctionSignatureHint`?

File: test/pkgs/Dynamo Samples/bin/SampleLibraryZeroTouch.Migrations.xml [hunk 2]
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0"?>
 <migrations>
   <priorNameHint>
-    <oldName>Examples.BasicExample.Create@Autodesk.DesignScript.Geometry.Point</oldName>
-    <newName>Examples.NEWBasicExample.Create@Autodesk.DesignScript.Geometry.Point</newName>
+    <oldName>Examples.BasicExample.Create</oldName>
+    <newName>Examples.NEWBasicExample.Create</newName>
```
Comment (aparajit-pratap) [inline test/pkgs/Dynamo Samples/bin/SampleLibraryZeroTouch.Migrations.xml:5]: Should we add another test with the parameter names included just to make sure that works as well? 
I think we should also add a test to test migration to a new method with the same name but with new parameters?

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs (hunk 1 above)
Comment (pinzart90) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs:233]: could simplify code here with coalescing operator :
`
var lookupSignature = libraryServices.FunctionSignatureFromFunctionSignatureHint(mangledName) ?? mangledName;
`
Unless your way is easier to read/debug.

File: test/pkgs/Dynamo Samples/bin/SampleLibraryZeroTouch.Migrations.xml (hunk 2 above)
Comment (mjkkirschner) [reply test/pkgs/Dynamo Samples/bin/SampleLibraryZeroTouch.Migrations.xml:5]: Theres an existing case below this one that tests your first requested case.

I'm not sure that your second requested case will actually work - if you change the parameters - this migration functionality has no way of wiring up the new parameters so they make sense... I can definitely try it out though and add a test.

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs (hunk 1 above)
Comment (mjkkirschner) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs:233]: can do.

Comment (mjkkirschner) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs:233]: will update.

File: test/pkgs/Dynamo Samples/bin/SampleLibraryZeroTouch.Migrations.xml (hunk 2 above)
Comment (aparajit-pratap) [reply test/pkgs/Dynamo Samples/bin/SampleLibraryZeroTouch.Migrations.xml:5]: What about cases like this: https://github.com/DynamoDS/Dynamo/blob/d1195bea43e6d6f0570670b5ed85d1bf723e9427/extern/ProtoGeometry/ProtoGeometry.Migrations.xml#L77?

Comment (aparajit-pratap) [reply test/pkgs/Dynamo Samples/bin/SampleLibraryZeroTouch.Migrations.xml:5]: And what about the case with the same namespace but different method name only?

Comment (mjkkirschner) [reply test/pkgs/Dynamo Samples/bin/SampleLibraryZeroTouch.Migrations.xml:5]: 1. I think this magically works by reflection right? The order of the parameters is unchanged and only curve as become curve[] ?  - you are interested in a test that checks for this but with a migration that does not include parameter types in the migration right?

2. I can definitely add that.

**Comment by mjkkirschner** [comment]
@aparajit-pratap I've added a new test case using FFITarget with a new class - it checks:

1. that a node with extra parameters is migrated correctly to a new method, with a new name - both methods still exist.
2. a method name is changed and the old method is missing. 
3. A node is migrated from one overload to another with the same name but a different number of parameters.

File: test/Engine/FFITarget/ClassWithMigratedMembers.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            return "migrated";
+        }
+
+        //we test a file that uses the node AMemberWithAChangedName0
+        public string AMemberWithAChangedName2(string a)
+        {
+            return a;
+        }
+      /*  public string AMemberWithAChangedName1(string a)
```
Comment (aparajit-pratap) [inline test/Engine/FFITarget/ClassWithMigratedMembers.cs:40]: Does this have to be commented out in order for the migration to work?

File: test/Engine/FFITarget/ClassWithMigratedMembers.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        public string AMemberWithTwoParams(string a, string b)
+        {
+            return a + b;
+        }
+        public string AMemberWithThreeParams(string a, string b,string c ="abc")
+        {
+            return "migrated";
+        }
+
+        //we test a file that uses the node AMemberWithAChangedName0
```
Comment (aparajit-pratap) [inline test/Engine/FFITarget/ClassWithMigratedMembers.cs:35]: I think you mean `AMemberWithAChangedName1`?

File: test/Engine/FFITarget/FFITarget.Migrations.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0"?>
+<migrations>
+  <priorNameHint>
+    <oldName>FFITarget.ClassWithMigratedMembers.AMemberWithAChangedName1</oldName>
+    <newName>FFITarget.ClassWithMigratedMembers.AMemberWithAChangedName2</newName>
+  </priorNameHint>
+  <priorNameHint>
+    <oldName>FFITarget.ClassWithMigratedMembers.AMemberWithTwoParams@string,string</oldName>
+    <newName>FFITarget.ClassWithMigratedMembers.AMemberWithThreeParams@string,string,string</newName>
```
Comment (aparajit-pratap) [inline test/Engine/FFITarget/FFITarget.Migrations.xml:9]: Does this work if you keep the method name the same?

File: test/Engine/FFITarget/ClassWithMigratedMembers.cs (hunk 3 above)
Comment (mjkkirschner) [reply test/Engine/FFITarget/ClassWithMigratedMembers.cs:40]: no, I just thought it was more explicit to illustrate what we were testing, a bit weird I admit.

File: test/Engine/FFITarget/ClassWithMigratedMembers.cs (hunk 4 above)
Comment (mjkkirschner) [reply test/Engine/FFITarget/ClassWithMigratedMembers.cs:35]: yes, thanks, will change.

**Review by aparajit-pratap — APPROVED** [review]

**Pull request merged**